### ISSUE TITLE
`useFetch` and `useAsyncData` type inference incorrect for generic parameter

### ISSUE THREAD

**Issue by miosenpai** [nuxt/nuxt#28030]
### Environment

```
------------------------------
- Operating System: Linux
- Node Version:     v18.20.3
- Nuxt Version:     3.12.3
- CLI Version:      3.12.0
- Nitro Version:    2.9.7
- Package Manager:  npm@10.2.3
- Builder:          -
- User Config:      devtools
- Runtime Modules:  -
- Build Modules:    -
------------------------------
```

### Reproduction

https://stackblitz.com/edit/github-8k6key?file=composables%2FuseCustomData.ts

### Describe the bug

When you use a generic parameter (e.g. the `T` in `<T>(...)`) to type the returned data of `useFetch` or `useAsyncData`, the functions end up inferring the returned data's type to be something with no fields in it (even if `T` was guaranteed to have certain fields via `extends`).

Other functions don't seem to have this problem (e.g. `$fetch` as demonstrated in the reproduction), so this behavior is unlikely to be caused by TypeScript itself (I think it might have something to do with how the `pick` option affects the type inference process).

**Comment by danielroe** [comment]
Yes, this is caused by the fact that it's not possible to know the keys of a type that extends another type.

We deliberately try to identify all possible keys (https://github.com/nuxt/framework/pull/9061), when passed a union (e.g. `{ title: string } | { description: string }`) not just the keys that are common to all possible types.

You can see a minimal example here:

https://www.typescriptlang.org/play/?#code/GYVwdgxgLglg9mABAUwB5WWAJgHgCoB8iAFAJQBciACjBANb4A0ieK6mWAzi4gPyJ1kATzjAelMMgBuyAE5EA3gChEqxLORQQspAoC+iAIbdDYIUr1LjQyInxsM2bgsQBbIQGlhlTlFkwwAHNEPQIyRABeRRU1CARfRCxDKENIhw58MNIAbhjVJJSAOncvcz1soA

It would be nice to have this just work in Nuxt but I think it's enough of an edge case that the best solution is for you resolve it yourself in the types of your custom composable.
